### PR TITLE
[ENH] Added TemporalConvolutionalNetwork in aeon/networks

### LINKED ISSUE #2931: [ENH] Implement the TemporalConvolutionalNetwork in aeon/networks
### Describe the feature or idea you want to propose

Introduce a robust and modular implementation of **Temporal Convolutional Networks (TCN)** for sequence modeling in `aeon.forecasting.deep_learning`. The architecture follows the foundational design proposed in the paper _"An Empirical Evaluation of Generic Convolutional and Recurrent Networks for Sequence Modeling"_ by Bai et al. (2018), and implements the TCN using Keras/TensorFlow, adapted from the original PyTorch version [locuslab/TCN](https://github.com/locuslab/TCN/blob/master/TCN/tcn.py).

The design is aimed at preserving the original model's key ideas:
- **Dilated convolutions** to exponentially expand the receptive field.
- **Residual connections** for improved stability during training.
- **Weight-normalized Conv1D** logic re-implemented through customized `Conv1D` layers in Keras.

### Describe your proposed solution

I have implemented this TCN architecture as a Keras-based class `TemporalConvolutionalNetwork`, inheriting from `BaseDeepLearningNetwork`. The following private methods abstract key functionalities of the model:

- `_conv1d_with_variable_padding(...)`: Handles variable causal padding logic akin to `padding + chomp` used in the Torch `Chomp1d` module.
- `_chomp_1d(...)`: Ensures strict causality by slicing excess padding.
- `_temporal_block(...)`: Builds individual residual blocks with two Conv1D layers, activations, dropout, and a residual connection with optional 1x1 projection.
- `_temporal_conv_net(...)`: Stacks temporal blocks with exponentially increasing dilation factors (i.e., `d=2^i`) to create a deep, memory-efficient model.
- `build_network(...)`: Orchestrates the overall architecture using the private components.

#### Example snippet (from `_tcn.py`)

```python
def _temporal_block(self, x, n_inputs, n_outputs, kernel_size, stride, dilation, padding, dropout, training):
    out = self._conv1d_with_variable_padding(x, n_outputs, kernel_size, padding, stride, dilation)
    out = self._chomp_1d(out, padding)
    ...
    if n_inputs != n_outputs:
        res = self._conv1d_with_variable_padding(x, n_outputs, 1, 0, 1, 1)
    result = tf.keras.layers.Add()([out, res])
    return tf.keras.layers.ReLU()(result)
```

The structure mirrors the reference PyTorch TemporalBlock:

```python
class TemporalBlock(nn.Module):
    ...
    self.conv1 = weight_norm(nn.Conv1d(...))
    self.chomp1 = Chomp1d(padding)
    ...
    self.net = nn.Sequential(...)
    self.downsample = nn.Conv1d(...) if n_inputs != n_outputs else None
```

### Describe alternatives you've considered, if relevant

N/A

### Additional context

N/A

### PR DESCRIPTION
**Pull request by lucifer4073**
#### Reference Issues/PRs

Fixes #2931 .

#### What does this implement/fix? Explain your changes.

Introduce a robust and modular implementation of **Temporal Convolutional Networks (TCN)** for sequence modeling in `aeon.forecasting.deep_learning`.

#### Does your contribution introduce a new dependency? If yes, which one?
N/A

#### Any other comments?

As of now, there is a weight normalization layer remaining. Need some clarifications before proceeding with that.

##### For all contributions
- [X] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [X] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/networks/__init__.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
     "AEBiGRUNetwork",
     "DisjointCNNNetwork",
     "RecurrentNetwork",
+    "TemporalConvolutionalNetwork",
```
Comment (hadifawaz1999) [inline aeon/networks/__init__.py]: i would prefer calling it `TCNNetwork` to keep consistency with others, and detail abbreviation in doctstring, like we do with `FCNNetwork` and `TimeCNNNetwork`

File: aeon/networks/_tcn.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
+"""Implementation of Temporal Convolutional Network (TCN).
+
+Based on the paper "An Empirical Evaluation of Generic Convolutional and
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: You can add a references section in docstring, check `InceptionNetwork` as example, don't put it in the header

File: aeon/networks/_tcn.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+    _config = {
+        "python_dependencies": ["tensorflow"],
+        "python_version": "<3.13",
+        "structure": "encoder",
+    }
+
+    def __init__(
+        self,
+        num_inputs: int = 1,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: as mentioned in the last meeting, you do not need this parameter, the number of input channels can be taken from the `input_shape` parameter of `build_network`

File: aeon/networks/_tcn.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+        Dropout rate applied after each convolutional layer for regularization.
+
+    Notes
+    -----
+    The receptive field size grows exponentially with network depth due to
+    dilated convolutions with dilation factors of 2^i for layer i.
+
+    References
+    ----------
+    Bai, S., Kolter, J. Z., & Koltun, V. (2018). An empirical evaluation of
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: add a number to the citation ad refer to it in the docstring description above (check other networks for example)

File: aeon/networks/_tcn.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    - Residual connections for training stability
+
+    The TCN can take sequences of any length and map them to output sequences
+    of the same length, making it suitable for autoregressive prediction tasks.
+
+    Parameters
+    ----------
+    num_inputs : int
+        Number of input channels/features in the input sequence.
+    num_channels : list of int
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: n_filters in docstring to follow the parameter name, no longer num_channels

File: aeon/networks/_tcn.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    _config = {
+        "python_dependencies": ["tensorflow"],
+        "python_version": "<3.13",
+        "structure": "encoder",
+    }
+
+    def __init__(
+        self,
+        num_inputs: int = 1,
+        n_filters: list = [16] * 3,  # changed from num_channels
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: you can remove the comment here

File: aeon/networks/_tcn.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        """
+        super().__init__()
+        self.num_inputs = num_inputs
+        self.n_filters = n_filters
+        self.kernel_size = kernel_size
+        self.dropout = dropout
+
+    def _conv1d_with_variable_padding(
+        self,
+        x,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: should be called `input_tensor` instead of `x`

File: aeon/networks/_tcn.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+            padding="valid",
+        )
+
+        # Apply convolution
+        out = conv_layer(padded_x)
+
+        # Transpose back to PyTorch format (batch, channels, sequence)
+        return tf.keras.layers.Permute((2, 1))(out)
+
+    def _chomp_1d(self, x, chomp_size: int):
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: can we call it something else than `chomp` ? am not understanding the nomenclature:p

Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `input_tensor` instead of `x`

File: aeon/networks/_tcn.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+        Returns
+        -------
+        tf.Tensor
+            Chomped tensor with reduced sequence length.
+        """
+        return x[:, :, :-chomp_size]
+
+    def _temporal_block(
+        self,
+        x,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `input_tensor` instead of `x`

File: aeon/networks/_tcn.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+        super().__init__()
+        self.num_inputs = num_inputs
+        self.n_filters = n_filters
+        self.kernel_size = kernel_size
+        self.dropout = dropout
+
+    def _conv1d_with_variable_padding(
+        self,
+        x,
+        filters: int,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `n_filters` instead of `filters`

File: aeon/networks/_tcn.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+        kernel_size: int,
+        padding_value: int,
+        stride: int = 1,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `strides` not `stride`

File: aeon/networks/_tcn.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+        """
+        return x[:, :, :-chomp_size]
+
+    def _temporal_block(
+        self,
+        x,
+        n_inputs: int,
+        n_outputs: int,
+        kernel_size: int,
+        stride: int,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `strides` not `stride`

File: aeon/networks/_tcn.py [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+        dilation: int,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `dilation_rate` not `dilation`

File: aeon/networks/_tcn.py [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+        tf.Tensor
+            Chomped tensor with reduced sequence length.
+        """
+        return x[:, :, :-chomp_size]
+
+    def _temporal_block(
+        self,
+        x,
+        n_inputs: int,
+        n_outputs: int,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `n_filters` not `n_outputs`

File: aeon/networks/_tcn.py [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+        kernel_size: int,
+        stride: int,
+        dilation: int,
+        padding: int,
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `padding_value` not `padding`

File: aeon/networks/_tcn.py [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+        out = self._conv1d_with_variable_padding(
+            out, n_outputs, kernel_size, padding, stride, dilation
+        )
+        out = self._chomp_1d(out, padding)
+        out = tf.keras.layers.ReLU()(out)
+        out = tf.keras.layers.Dropout(dropout)(out, training=training)
+
+        # Residual connection with optional dimension matching
+        if n_inputs != n_outputs:
+            res = self._conv1d_with_variable_padding(x, n_outputs, 1, 0, 1, 1)
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: can you specify the parameters being set here ? it would more clear instead of just 1,0,1,1

File: aeon/networks/_tcn.py [hunk 17]
```diff
@@ -0,0 +1,341 @@
...
+            Whether the model is in training mode.
+
+        Returns
+        -------
+        tf.Tensor
+            Output tensor after applying all temporal blocks.
+        """
+        num_levels = len(n_filters)
+        for i in range(num_levels):
+            dilation_size = 2**i
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `dilation_rate`

File: aeon/networks/_tcn.py [hunk 18]
```diff
@@ -0,0 +1,341 @@
...
+    def build_network(self, input_shape: tuple, **kwargs) -> tuple:
+        """Build the complete TCN architecture.
+
+        Constructs a series of temporal blocks with exponentially increasing
+        dilation factors to achieve a large receptive field efficiently.
+
+        Parameters
+        ----------
+        input_shape : tuple
+            Shape of input data (sequence_length, num_features).
```
Comment (hadifawaz1999) [inline aeon/networks/_tcn.py]: `n_timepoints` and `n_channels` instead

File: aeon/networks/_tcn.py (hunk 2 above)
Comment (lucifer4073) [reply aeon/networks/_tcn.py]: Already there in base class docstring

File: aeon/networks/_tcn.py (hunk 5 above)
Comment (lucifer4073) [reply aeon/networks/_tcn.py]: n_filters name is failing, because `aeon/networks/tests/test_all_networks.py` expects n_filters to be int and not list

Comment (lucifer4073) [reply aeon/networks/_tcn.py]: I plan to use n_blocks instead.

**Comment by lucifer4073** [comment]
Hi @hadifawaz1999,

Requested changes have been made.

**Comment by lucifer4073** [comment]
Hi @hadifawaz1999 ,

Request changes have been made to TCN including changing input shape to `(n_timpoints, n_channels)`

**Review by hadifawaz1999 — APPROVED** [review]
